### PR TITLE
OUT-243 - API integration for dynamic workflow states and tasks

### PR DESCRIPTION
**Pull request by aatbip**
### Tasks

- [API Integration for dynamic workflow states and tasks](https://linear.app/copilotplatforms/issue/OUT-243/api-integration-for-dynamic-status-columns)

### Changes

- [x] Adds global error handling file in the root
- [x] Adds TaskResponse and WorkflowStateResponse types 
- [x] Adds HOC to handle client side state update
- [x] Adds util function to encode and decode string for URL params
- [x] Adds function in server component to fetch workflowState and tasks
- [x] Renders workflow state and tasks dynamically 
- [x] Refactors drag and drop to fit the responses coming from the API

### Testing Criteria

-  [Loom](https://www.loom.com/share/a5431c2a9a594fd38f2e0e67366a8546?sid=2e5e9fac-bf58-40a5-95a6-3bca0c5abc95)

### Notes

- This PR is dependent to #14. I have merged the latest of #14 with this PR.

### CODE REVIEW

File: src/app/page.tsx
Comment (rrojan) [inline src/app/page.tsx:10]: Please remove the `apiUrl` part since its not needed

File: src/app/ui/TaskBoard.tsx
Comment (rrojan) [inline src/app/ui/TaskBoard.tsx:64]: Why are we using `task.title` as a dynamic param? Shouldn't `task.id` be enough? Task title will be fetched from the API anyways.

File: src/components/cards/ClientTaskCard.tsx
Comment (rrojan) [inline src/components/cards/ClientTaskCard.tsx:4]: Please configure your prettier config to prefer absolute / aliased imports

File: src/hoc/DragDropHandler.tsx
Comment (rrojan) [inline src/hoc/DragDropHandler.tsx:37]: Please remove commented code if not needed

File: src/hoc/DragDropHandler.tsx
Comment (rrojan) [inline src/hoc/DragDropHandler.tsx:52]: It would be better to make this a custom type. I see this is used at multiple places so giving it a meaningful type name would also make it easier to understand

File: src/hoc/DragDropHandler.tsx
Comment (rrojan) [inline src/hoc/DragDropHandler.tsx:63]:
```tsx
style={{ opacity }}
```

File: src/redux/store.ts
Comment (rrojan) [inline src/redux/store.ts:3]: Please prefer aliased imports

File: src/types/dto/tasks.dto.ts
Comment (rrojan) [inline src/types/dto/tasks.dto.ts:36]: Please merge `main` into this branch once to get rid of all API related changes that are still present here as new changes

File: src/utils/generateParamString.ts
Comment (rrojan) [inline src/utils/generateParamString.ts:5]: Once we remove `task.title` from our URL this should no longer be required

**Review by rrojan — CHANGES_REQUESTED** [review]

File: src/app/ui/TaskBoard.tsx
Comment (aatbip) [reply src/app/ui/TaskBoard.tsx:64]: In the server rendered apps, we try to get rid of the loading states as much as possible by fetching the data in the server and pre-rendering the parts of the page whenever possible. 
This is also an approach on pre-rendering the `title` in the details page even before the data is fetched. By passing the `title` through params in the server component ([link to code](https://github.com/copilot-platforms/tasks-app/blob/063f90c465e8b03d3f05b314cff12a11ddda8ec2/src/app/detail/%5Btask_id%5D/%5Btask_name%5D/%5Buser_type%5D/page.tsx#L20)), we can pre-render this part and show the users with something. This is an approach where we take benefits of server side rendering by displaying parts of pages without data fetching is completed. It is also a composition pattern for designing apps using mixed components. 

Beside this, I think putting `title` along with `id` as part of a route param is a tradition. For instance, this is an example of the linear details page route: `https://linear.app/copilotplatforms/issue/OUT-249/implement-remaining-task-actions`

File: src/hoc/DragDropHandler.tsx
Comment (aatbip) [reply src/hoc/DragDropHandler.tsx:52]: I couldn't understand properly where should the type be given. Can you please explain? @rrojan

File: src/hoc/DragDropHandler.tsx
Comment (aatbip) [reply src/hoc/DragDropHandler.tsx:63]: Changed in commit b5d2a564e35db923a74b0c59da764300ce1a9193

File: src/types/dto/tasks.dto.ts
Comment (aatbip) [reply src/types/dto/tasks.dto.ts:36]: Merged latest main to this branch

File: src/utils/generateParamString.ts
Comment (aatbip) [reply src/utils/generateParamString.ts:5]: I think having `task.title` in the route param is a good approach for the reasons mentions [here](https://github.com/copilot-platforms/tasks-app/pull/15#discussion_r1577753143). 

If there are any potential reasons for this to be removed then please do let me know thoughts.

File: src/redux/store.ts
Comment (aatbip) [reply src/redux/store.ts:3]: I think we should setup prettier config for aliased import in this app. Better to do this in a separate PR I guess.

File: src/app/ui/TaskBoard.tsx
Comment (rrojan) [reply src/app/ui/TaskBoard.tsx:64]: I see that's what Linear does as well.. although it seems weird at first because of the long URLs... in a later PR we could make it so that links like `tasks.app/WEB-01` redirects to `tasks.app/WEB-01/Do-something` as well (like Linear). Right now the `WEB-01` part is hardcoded so it might not be possible.

File: src/redux/store.ts
Comment (rrojan) [reply src/redux/store.ts:3]: Yes... it requires an extension and some hit-and-trial. I have it in my VSCode settings - will move it to prettierrc in a later PR :raised_hands:

File: src/app/ui/TaskBoard.tsx
Comment (rrojan) [reply src/app/ui/TaskBoard.tsx:64]: There shouldn't be much issues if it is only used this way. However there might be a case where I have opened "tasks.app/WEB-01/Do-something" in my browser and the author of the ticket renames it from "Do something" to "Do something else".

I that case can you confirm if "Do something else" is shown as title when "tasks.app/WEB-01/Do-something" is opened? (Even if "Do-something" is used to set the initial title, the title is updated in place if necessary?

Let me know if this is confusing :bow:

**Review by rrojan — DISMISSED** [review]
LGTM :trophy: 

Please look into the title comment above though and see if that is a necessary consideration.

File: src/app/ui/TaskBoard.tsx
Comment (aatbip) [reply src/app/ui/TaskBoard.tsx:64]: Good catch. The synchronization in the case title is changed was not implemented. Considering this and also considering the long URL until we have shorter task id like `WEB-01`, I have removed the task name from route param in d69c36b75be57aa82c305ceefd656dec1264f264 commit. 

In the future, we can have the title in route param by managing the synchronization in the case if title is changed, if required. Otherwise, this is fine. @rrojan

**Comment by aatbip** [comment]
Waiting for #10 to get merged before merging this PR. 
@rrojan

File: src/hoc/ClientSideStateUpdate.tsx
Comment (makeitraina) [inline src/hoc/ClientSideStateUpdate.tsx:16]: You have the right idea about this store but it requires a little more setup. Please follow this guide: https://redux-toolkit.js.org/usage/nextjs#introduction

**Review by makeitraina — COMMENTED** [review]
Nice work on getting many of the different pieces connected here Ananta.

This PR is quite big and covers SEVERAL different requirements and setup. I did my best to give a holistic pass on this and commented on several areas. Let me know how you want to proceed here.

When can either tackle these improvements separately or try to rebuild this PR with incremental changes. If we choose to do it separately, I think we should have a plan on what the individual parts are and agree on the approach. It might actually be easier to write a separate doc with how to cover feedback more like a TDD.

File: src/app/page.tsx
Comment (aatbip) [reply src/app/page.tsx:52]: We haven't created the tasks for real-time updates already. This feature will come in the next sprint. 

The approach is as follows:
- For the first render, tasks will be fetched in the server side so that we can also leverage SSR on the first render. This is also important because we are using `revalidation` i.e. `revalidateTag('getAllTasks')` using server actions during mutations, for ex [HERE](https://github.com/copilot-platforms/tasks-app/blob/fb44de20a05ad6833685408c12bccc623cb5b052/src/app/actions.ts#L22).
- After the first render, we can make use of SWR or similar to implement real-time updates using polling by fetching the data in intervals from the client side.

File: src/app/error.tsx
Comment (aatbip) [reply src/app/error.tsx:1]: According to the next.js doc [HERE](https://nextjs.org/docs/app/building-your-application/routing/error-handling), it should be a client component. The reason is it is useful for catching unexpected errors that occur in both Server Components and Client Components.

File: src/app/actions.ts
Comment (aatbip) [reply src/app/actions.ts]: Updated in 220ff7ae81dcc75235449ca506048cc14d235d0c commit.

File: src/app/actions.ts
Comment (aatbip) [reply src/app/actions.ts:10]: `revalidateTag` will revalidate the data for the fetch functions that matches the specified tags. For instance, `getAllTasks` tag is assigned to [THIS](https://github.com/copilot-platforms/tasks-app/blob/220ff7ae81dcc75235449ca506048cc14d235d0c/src/app/page.tsx#L26) fetch.

File: src/app/page.tsx
Comment (aatbip) [reply src/app/page.tsx]: Updated in 6e8f7f4385aeb454f95b67bd46919860274ba2bc commit

File: src/app/ui/TaskBoard.tsx
Comment (aatbip) [reply src/app/ui/TaskBoard.tsx]: Do you mean the type of `handleCreate`? If so then type of `handleCreate` varies in between `TaskBoard` and `NewTaskForm` components. 
Also, `updateWorkflowStateIdOfTask` is depreciated in #23 PR. Please check [HERE](https://github.com/copilot-platforms/tasks-app/blob/3a1d7d40019db46f168f9ff7c82d1714c5a6393f/src/app/actions.ts#L26).

File: src/app/ui/TaskBoard.tsx
Comment (aatbip) [reply src/app/ui/TaskBoard.tsx:32]: The `updateWorlflowStateIdByTaskId` reducer is responsible for updating the `workflowStateId` of the task optimistically. This is useful for handling the drag and drop functionality. This is not intended here to perform mutation and update the store with the new data after mutation. However, this part is being handled using the `revalidateTag` approach currently.

File: src/components/cards/ClientTaskCard.tsx
Comment (aatbip) [reply src/components/cards/ClientTaskCard.tsx:23]: `WEB-01` is mock representation for the unique ID. This component has been refactored in #23.

File: src/components/inputs/Selector.tsx
Comment (aatbip) [reply src/components/inputs/Selector.tsx:34]: Typecasting is done here to cast the `unknown` type for `option`. We can make use to typeguard as it is more safer approach and I think it is better to do this in a separate PR to avoid conflicts. This has to be changed in multiple files.

File: src/components/inputs/Selector.tsx
Comment (aatbip) [reply src/components/inputs/Selector.tsx:142]: Please refer to [this comment](https://github.com/copilot-platforms/tasks-app/pull/15#discussion_r1584564860)

File: src/hoc/ClientSideStateUpdate.tsx
Comment (aatbip) [reply src/hoc/ClientSideStateUpdate.tsx:16]: The redux setup I have done here is already following the doc you've shared above. Can you please let me know if there is anything specific you would like me to do regarding the redux setup? 

Also, this `ClickSideStateUpdate` is not a store but an HOC. It is just responsible for updating the redux store in the client side. This component is a client component and as we can't update redux store from the server component, I am creating this HOC to update the store if necessary. Please let me know your thoughts on this.

**Review by rrojan — DISMISSED** [review]
LGTM :trophy:

**Review by rrojan — DISMISSED** [review]
LGTM :trophy:

**Review by rrojan — APPROVED** [review]
LGTM again :trophy:

**Pull request merged**
